aip/0135.md

```diff
@@ -179,6 +179,30 @@ message DeleteBookRequest {
 If the etag is provided and does not match the server-computed etag, the
 request **must** fail with a `FAILED_PRECONDITION` error code.
 
+### Deletion prevention
+
+Sometimes, it may be desirable for users to protect a resource against accidental
+deletion. This is often especially useful in "infrastructure as code" deployments
+where resources are kept in sync with some declarative configuration.
+
+To achieve this, the resource **should** include a `bool prevent_delete` field to
+provide opt-in protection. This field **may** be added retroactively to existing
+resources.
+
+```proto
+message Book {
+  // The name of the book.
+  // Format: publishers/{publisher}/books/{book}
+  string name = 1;
+
+  // Optional. While true, this Book cannot be deleted.
+  bool prevent_delete = 2;
```

Comment:
A single bit like this leave a lot to be desired, in particular:
  - It does not convey _why_ a resource should not be deleted
  - It does not allow the expression of multiple delete-blocking conditions

In general, I argue that the behavior of blocking deletion is better handled as a policy attached to a resource, as opposed to a property of the resource itself.

An example of such a policy is Liens in the GCP CloudResourceManager API. http://cloud/resource-manager/docs/project-liens

While more verbose than a single bit, these policies allow the expression of multiple reasons to not delete a Project, affording those reasons to come and go at various points in time (with appropriate audit logging).

I believe within the next few months there will be other examples of polices which can achieve this goal (and decline to comment further in this forum).

Reply:
I'm in agreement here. While this seems like it fixes the immediate issue, I'm concerned that it is a very narrow solution that prevents lots of potential future use cases....

Reply:
This is interesting prior art. Do we need to block until this is more fleshed out? Or do we think we could get away with something extensible? For example:

message Book {
  ...
  message Lien {
    string origin = 1;
    string reason = 2;
    // Additional options like expiration or IAM selectors can be added as needed
  }
  repeated Lien liens = X;
}

Alternatively, we could make Liens a subresource, which simplifies patching but might complicate discovery.

Reply:
Policies are not represented in the resources themselves, so you would never have a Lien be a field or a subresource. Policies are _attached_ or _bound_ to resources.

Liens are a type of policy. [IAM](http://cloud/iam/docs/overview) and [OrganizationPolicy](http://cloud/resource-manager/docs/organization-policy/creating-managing-policies) are other examples of policies.

Personally I think it would be best to wait on this for a few months until upcoming features surface.

Reply:
Hmm, I wonder if this is getting a bit heavyweight in that case? We're getting into an entirely separate resource hierarchy, and potentially something that ought to be made into a common API.

Something like this certainly makes sense for high-level containers like projects. But it seems a bit less necessary for "leaf" resources like databases. At that level, the reason you don't want to delete is pretty clear: you don't want to break production!

Is it possible that we want two options here, aka "simple" and "extensible"?

Reply:
> Something like this certainly makes sense for high-level containers like projects. But it seems a bit less necessary for "leaf" resources like databases. At that level, the reason you don't want to delete is pretty clear: you don't want to break production!

We can presumably handle that in the IAM system though.

Reply:
I might be misunderstanding, but wanted to call something out just in case:

We *can* handle basically the same functionality as this "flag" with access control, but having permission to delete something and preventing accidental deletion are very different. 

GCS has an implicit lock on buckets (to delete a bucket you first have to delete all objects). I might want to delete something, but I also want to ensure that there's an extra step in there for anyone who wants to delete something that forces them to double check before the thing disappears. Think back to cassette tapes where you had the tab that stopped you from taping over something you recorded.

IAM is permission. This flag is "double checking". I think they're related but not the same thing.

Reply:
Thanks for the feedback everyone.

tl;dr: If possible, I would like to **quickly** land in one of the following worlds:
- We agree that a simple bit is an option under some circumstances (we can work these out later), and we agree on the **name**
- We agree that deletion protection will **never** be offered in the form of a bit

Some background will be useful here. Terraform has developed a nasty habit of deleting/recreating Bigtable tables and instances. This is because the behavior of the framework is to "force new" resources to match the user's declarative config, unless code has been explicitly written in the Terraform provider to do an in place update.

The Terraform framework has a way to prevent this behavior, but it's turned off by default and many users aren't aware of it. We are pushing on Hashicorp to improve this, but it's a large change for them and will likely take a long time. Users also shouldn't be using the auto-apply functionality that tends to cause these problems, but in practice we've seen that they often do. So given that we've seen multiple incidents of accidental data loss from this issue, we would like to add some form of **default on** protection ASAP.

Our working design for this is to add a new field to the Bigtable Terraform config which offers deletion protection. Terraform can make this field default on, and can enforce it client-side without a corresponding API field. Users upgrading to a new version of the Terraform provider will need to explicitly disable this protection before deleting a resource. This bit will likely be useful for other databases as well, so I've been thinking of it as a new "standard."

Because it can start client-side, we can add this field in a vacuum if needed. But it feels like we ought to think about how it will figure into future API integrations. In particular, if Bigtable and others later decide to add API-level protection, it would make sense for setting the Terraform bit to turn that protection on. And in that case it would be weird if the Terraform field and the API field were named differently.

That said, it may not be that weird if the API-level mechanism is qualitatively different. So if setting this bit in Terraform will eventually translate to creating a Lien in the API (origin=Terraform, reason=prevent_delete), the naming issue goes away. In that case we would probably call it deletion_protection to match the existing GCE API. Note that we can always add a "Lien" resource to Terraform later, so adding a bit first doesn't constrain our options like an AIP does.

Due to the severity of the underlying issue, I don't want to block the new Terraform field any longer than necessary. The full AIP isn't blocking, but figuring out any constraints on the field name is. If the consensus is that APIs should **never** use a simple bit, then I think we're fully decoupled. But I'd like to confirm this (soon) before proceeding.

Reply:
Ping! It's been one week. If I don't get a response soon I will tell Terraform to go ahead with the name "deletion_protection"

Reply:
We have gone ahead with this approach

Reply:
Sorry for failing to respond. I think `deletion_protection` for Terraform is fine (and I assume you know Terraform style).